lib/Command/CleanUp.php

```diff
@@ -0,0 +1,115 @@
+ * it under the terms of the GNU Affero General Public License, version 3,
+ * as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License, version 3,
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>
+ *
+ */
+
+namespace OCA\TwoFactorWebauthn\Command;
+
+use OCA\TwoFactorWebauthn\Db\PublicKeyCredentialEntityMapper;
+use OCP\DB\Exception;
+use OCP\IDBConnection;
+use OCP\IUserManager;
+use Symfony\Component\Console\Command\Command;
+use Symfony\Component\Console\Input\InputInterface;
+use Symfony\Component\Console\Output\OutputInterface;
+use Symfony\Component\Console\Style\SymfonyStyle;
+
+class CleanUp extends Command {
+	/** @var IDBConnection */
+	private $db;
+
+	/** @var IUserManager */
+	private $userManager;
+
+	/** @var PublicKeyCredentialEntityMapper */
+	private $webauthnMapper;
+
+	public function __construct(
+		IDBConnection $db,
+		IUserManager $userManager,
+		PublicKeyCredentialEntityMapper $webauthnMapper
+	) {
+		parent::__construct();
+
+		$this->db = $db;
+		$this->userManager = $userManager;
+		$this->webauthnMapper = $webauthnMapper;
+	}
+
+	protected function configure(): void {
+		$this->setName('twofactor_webauthn:cleanup');
+	}
+
+	protected function execute(InputInterface $input, OutputInterface $output): int {
+		$io = new SymfonyStyle($input, $output);
+		$io->title('Find orphaned webauthn credentials');
+		$io->text([
+			'This command search webauthn credentials for deleted users.',
+			'Found credentials for deleted users will be removed.'
+		]);
+
+		$confirm = $io->confirm('Start process?', true);
+		if ($confirm === false) {
+			$io->text('Thank you for using Two-Factor WebAuth!');
+			return 0;
+		}
+
+		foreach ($this->findUserIds() as $userId) {
+			if ($this->userManager->userExists($userId) === false) {
+				try {
+					$io->text('Delete credentials for uid "' . $userId . '"');
+					$this->webauthnMapper->deletePublicKeyCredentialsByUserId($userId);
+				} catch (Exception $e) {
+					$io->caution('Error deleting credentials: ' . $e->getMessage());
+				}
+			}
+		}
+
+		$io->success('Orphaned webauthn credentials removed.');
+
+		$io->text('Thank you for using Two-Factor WebAuth!');
+		return 0;
+	}
+
+	private function findUserIds(): array {
+		$userIds = [];
+
+		$qb = $this->db->getQueryBuilder()
+			->selectDistinct('user_handle')
+			->from('twofactor_webauthn_regs');
+
+		try {
+			$result = $qb->executeQuery();
+
+			while ($row = $result->fetch()) {
+				$userIds[] = $row['user_handle'];
+			}
+
+			$result->closeCursor();
+		} catch (\Exception $exception) {
+			// what could go wrong ;)
```

Comment:
I'd just fail hard. There is no point in continuing

Reply:
OK